nfsuserver/nfsuserver2.cpp

```diff
@@ -63,10 +66,7135 @@ bool running = true;
+
+bool sort_REP_All(PlayerStat a, PlayerStat b) {
+	return a.REP_All > b.REP_All;
+};
+
+bool sort_REP_Circ(PlayerStat a, PlayerStat b) {
+	return a.REP_Circ > b.REP_Circ;
+};
+
+bool sort_REP_Sprint(PlayerStat a, PlayerStat b) {
+	return a.REP_Sprint > b.REP_Sprint;
+};
+
+bool sort_REP_Drag(PlayerStat a, PlayerStat b) {
+	return a.REP_Drag > b.REP_Drag;
+};
+
+bool sort_REP_Drift(PlayerStat a, PlayerStat b) {
+	return a.REP_Drift > b.REP_Drift;
+};
+
+bool sort_Time(StarsLap a, StarsLap b) {
+	return a.Time < b.Time;
+};
+
+bool sort_Points(StarsDrift a, StarsDrift b) {
+	return a.Points > b.Points;
+};
+
+
+#define NFSU_LAN_VERSION "2.0.1"
+#define DEFAULT_NEWS "-=-=-=-\nDefault news\nPlz tell server admin to make news.txt file ;)\n-=-=-=-=-"
+
+ServerClass Server; //core ;)
+
+SessionsClass Sessions; // started races players info
+
+char ascii[256] =
+{
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 62, 64, 64, 64, 63,
+  52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 64, 64, 64, 64, 64, 64,
+  64, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14,
+  15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64, 64,
+  64, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
+  41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
+};
+
+void base64_out(char* buf, char* obuf, int len) {
+	int nprbytes;
+	char* p = buf;
+	while (ascii[(int)*(p++)] <= 63);
+
+	nprbytes = len - 1;
+
+	while (nprbytes > 4 && *buf != '\0') {
+		*(obuf++) = (ascii[(int)*buf] << 2 | ascii[(int)buf[1]] >> 4);
+		*(obuf++) = (ascii[(int)buf[1]] << 4 | ascii[(int)buf[2]] >> 2);
+		*(obuf++) = (ascii[(int)buf[2]] << 6 | ascii[(int)buf[3]]);
+		buf += 4;
+		nprbytes -= 4;
+	}
+	if (nprbytes > 1)
+		*(obuf++) = (ascii[(int)*buf] << 2 | ascii[(int)buf[1]] >> 4);
+	if (nprbytes > 2)
+		*(obuf++) = (ascii[(int)buf[1]] << 4 | ascii[(int)buf[2]] >> 2);
+	if (nprbytes > 3)
+		*(obuf++) = (ascii[(int)buf[2]] << 6 | ascii[(int)buf[3]]);
+	*(obuf)++ = '\0';
+};
+
+void ShiftPersona(UserClass* user, RegUser* tr, int l) {
+	strcpy(user->Personas[l], user->Personas[l + 1]);
+	memset(user->Personas[l + 1], 0, 16);
+	strcpy(tr->Personas[l], tr->Personas[l + 1]);
+	memset(tr->Personas[l + 1], 0, 16);
+};
+
+bool IsCheater(const unsigned char wei, const unsigned char sus, const unsigned char eng, const unsigned char tur,
```

Comment:
Could you describe in simple words how this function works?

Reply:
I'll try to be briefly.
A player can get only 3 TJ's parts of 9 when completing a career mode. This function determines whether a set of TJ's parts is valid if a player had complete the career. The player is a cheater when:

- more than 3 TJ's parts
- doesn't have one of first set of TJ's parts (engine, transmission, tires)
- has more than one of the third set of TJ's pars (weight reduction, suspension, NOS)